src/Microsoft.Extensions.Http.Polly/DependencyInjection/PollyHttpClientBuilderExtensions.cs

```diff
@@ -0,0 +1,69 @@
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Net;
+using System.Net.Http;
+using Microsoft.Extensions.Http;
+using Polly;
+
+namespace Microsoft.Extensions.DependencyInjection
+{
+    public static class PollyHttpClientBuilderExtensions
+    {
+        public static IHttpClientBuilder AddPolicyHandler(this IHttpClientBuilder builder, IAsyncPolicy policy)
+        {
+            //builder.AddHttpMessageHandler(() => new PolicyHttpMessageHandler(createPolicy()));
+            return builder;
+        }
+
+        public static IHttpClientBuilder AddPolicyHandler(this IHttpClientBuilder builder, IAsyncPolicy<HttpResponseMessage> policy)
+        {
+            //builder.AddHttpMessageHandler(() => new PolicyHttpMessageHandler(createPolicy()));
+            return builder;
+        }
+
+        public static IHttpClientBuilder AddExceptionPolicyHandler(this IHttpClientBuilder builder, Func<PolicyBuilder, IAsyncPolicy> createPolicy)
+        {
+            var policyBuilder = Policy.Handle<HttpRequestException>();
+            var policy = createPolicy(policyBuilder);
+
+            //builder.AddHttpMessageHandler(() => new PolicyHttpMessageHandler(createPolicy()));
+            return builder;
+        }
+
+        public static IHttpClientBuilder AddExceptionPolicyHandler(this IHttpClientBuilder builder, Func<PolicyBuilder, IAsyncPolicy<HttpResponseMessage>> createPolicy)
+        {
+            //builder.AddHttpMessageHandler(() => new PolicyHttpMessageHandler(createPolicy()));
+            return builder;
+        }
+
+        public static IHttpClientBuilder AddServerErrorPolicyHandler(
+            this IHttpClientBuilder builder, 
+            Func<PolicyBuilder<HttpResponseMessage>, IAsyncPolicy<HttpResponseMessage>> createPolicy)
+        {
+            var policyBuilder = Policy.Handle<HttpRequestException>().OrResult<HttpResponseMessage>(response =>
+            {
+                return response.StatusCode >= HttpStatusCode.InternalServerError;
+            });
+            var policy = createPolicy(policyBuilder);
+
+            //builder.AddHttpMessageHandler(() => new PolicyHttpMessageHandler(createPolicy()));
+            return builder;
+        }
+
+        public static IHttpClientBuilder AddBadRequestPolicyHandler(
+            this IHttpClientBuilder builder,
+            Func<PolicyBuilder<HttpResponseMessage>, IAsyncPolicy<HttpResponseMessage>> createPolicy)
+        {
+            var policyBuilder = Policy.Handle<HttpRequestException>().OrResult<HttpResponseMessage>(response =>
+            {
+                return response.StatusCode >= HttpStatusCode.BadRequest;
+            });
+            var policy = createPolicy(policyBuilder);
+
+            //builder.AddHttpMessageHandler(() => new PolicyHttpMessageHandler(createPolicy()));
+            return builder;
+        }
+    }
```

Comment:
Is there as clear motivation for an `AddBadRequestPolicyHandler()` covering the 4xx codes generally, as for the other overloads?  Where Http status codes in the 4xx range indicate client errors, there may be no point in retrying these (eg if a request is badly formed 400, resubmitting it is usually not going to make it succeed).  Likewise, one typically might not want to take client-errors as cause to break a circuit-breaker.

Some specific exceptions: 

+ 408 (afaiu) the server signalling giving up waiting for the client's request to arrive complete; it could reflect wider network connectivity which merits a retry.
+ 429 RetryAfter: Polly has retry overloads which can handle this case (extract how long to wait before the next try, from the 429 response), but it remains a fairly specific configuration.

Reply:
> Is there as clear motivation for an AddBadRequestPolicyHandler() covering the 4xx codes generally, as for the other overloads?

Not really. That was just a straw man for us to beat up. 

I still want to get more feedback on whether we should have both `AddExceptionPolicyHandler` and `AddServerErrorPolicyHandler`.. When might you want to apply a policy to a connection failure but not a 500? Still needs more thought IMO.

Reply:
>[should we] have both AddExceptionPolicyHandler and AddServerErrorPolicyHandler?
>When might you want to apply a policy to a connection failure but not a 500? 

If the questions asked by users to the Polly team are a relevant measure, then handling them differently seems at least uncommon.  We received a lot of requests for policies to handle result codes as well as exceptions, but since adding that, no requests for help how to handle them differently.  My feeling (fwiw) is the majority use case defines a common strategy for all kinds of transient fault at an endpoint.  

If `AddExceptionPolicyHandler` and `AddServerErrorPolicyHandler` were merged, the power overloads you are proposing would still provide, for those who want more nuanced strategies.

Still interested what others may think.

---

The two enquiries the Polly team handles _periodically_ (I would not say frequently) on behaviour for specific result codes are:

+ 401 Unauthorized: [Can I make a policy to re-authenticate, then retry, on 401 Unauthorised?](https://github.com/App-vNext/Polly/wiki/Retry#retry-to-refresh-authorization)
+ 429 RetryAfter: [How do I make a wait-and-retry policy wait for a duration specified in an `x-ms-retry-after-ms` return header?](https://github.com/App-vNext/Polly/wiki/Retry#retryafter-when-the-response-specifies-how-long-to-wait)

We also in the Polly doco [recommend retrying on 408s](https://github.com/App-vNext/Polly/#step-1b-optionally-specify-return-results-you-want-to-handle), similar to [advice from the Azure team](https://docs.microsoft.com/en-us/azure/architecture/best-practices/retry-service-specific#general-rest-and-retry-guidelines).

Reply:
I think it's worth considering if we want to follow up on a convenience method for 'retryable' faults. Adding it to my todo list.

Regarding AuthZ, we plan to do that in the future (not this release). I'm not sure that plan is related to Polly, but we're interested in your thoughts as well.

Reply:
>I think it's worth considering if we want to follow up on a convenience method for 'retryable' faults. 

Happy to hear more what is the thinking here.  Is this around helpers for commoner cases of _how_ to handle, or something else?

Reply:
> Happy to hear more what is the thinking here. Is this around helpers for commoner cases of how one might handle (common retry strategies such as exponential backoff), or something else?

I think we'd want to provide a more opinionated approach to *what we consider a 'retry-able' request*. The starting point would be the info that you've already shared with us from your docs/guidance.

I think the polly API for creating different retry policies is already sufficient. 

If you have more ideas about this, feel free to open an issue for more focused discussion. Your feedback has been really valuable.